src/loop_libraries/library_collier.cpp

```diff
@@ -0,0 +1,167 @@
+// or (at your option) any later version.
+//
+// FlexibleSUSY is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with FlexibleSUSY.  If not, see
+// <http://www.gnu.org/licenses/>.
+// ====================================================================
+
+#include <limits>
+#include <boost/preprocessor/cat.hpp>
+#include <boost/preprocessor/repetition/repeat.hpp>
+#include "fortran_utils.hpp"
+#include "library_collier.hpp"
+
+#define COLLIER_TYPE(Z, N, TEXT) const std::complex<double>*,
+#define COLLIER_ARGS(N) BOOST_PP_REPEAT(N,COLLIER_TYPE,) const std::complex<double>*
+
+#define IMPL(R,ARGS,NAME) std::complex<double> BOOST_PP_CAT(NAME,_impl)(COLLIER_ARGS(ARGS));
+
+/* Non-vanishing imaginary parts of momentum invariants are not yet
+* suppoted by the current version (1.2.4) of COLLIER. */
+#define COLLIER_B(R,ARGS,NAME) std::complex<double> Collier::NAME ARGS noexcept\
+{\
+   const std::complex<double> p10 (p10_in.real(), 0.);\
+   const std::complex<double> m02 = m02_in;\
+   const std::complex<double> m12 = m12_in;\
+\
+   set_mu2_uv(scl2_in);\
+   return BOOST_PP_CAT(NAME,_impl)(&p10, &m02, &m12);\
+}
+#define COLLIER_C(R,ARGS,NAME) std::complex<double> Collier::NAME ARGS noexcept\
+{\
+   const std::complex<double> p10 (p10_in.real(), 0.);\
+   const std::complex<double> p21 (p21_in.real(), 0.);\
+   const std::complex<double> p20 (p20_in.real(), 0.);\
+   const std::complex<double> m02 = m02_in;\
+   const std::complex<double> m12 = m12_in;\
+   const std::complex<double> m22 = m22_in;\
+\
+   set_mu2_uv(scl2_in);\
+   return BOOST_PP_CAT(NAME,_impl)(&p10, &p21, &p20, &m02, &m12, &m22);\
+}
+#define COLLIER_D(R,ARGS,NAME) std::complex<double> Collier::NAME ARGS noexcept\
+{\
+   const std::complex<double> p10 (p10_in.real(), 0.);\
+   const std::complex<double> p21 (p21_in.real(), 0.);\
+   const std::complex<double> p32 (p32_in.real(), 0.);\
+   const std::complex<double> p30 (p30_in.real(), 0.);\
+   const std::complex<double> p20 (p20_in.real(), 0.);\
+   const std::complex<double> p31 (p31_in.real(), 0.);\
+   const std::complex<double> m02 = m02_in;\
+   const std::complex<double> m12 = m12_in;\
+   const std::complex<double> m22 = m22_in;\
+   const std::complex<double> m32 = m32_in;\
+\
+   set_mu2_uv(scl2_in);\
+   return BOOST_PP_CAT(NAME,_impl)(&p10, &p21, &p32, &p30, &p20, &p31, &m02, &m12, &m22, &m32);\
+}
+
+// Fortran wrapper routines
+extern "C" {
+   void initialize_collier_impl();
+   void set_mu2_uv_impl(double*);
+
+   BOOST_PP_SEQ_FOR_EACH(IMPL,0,A_SEQ)
+   BOOST_PP_SEQ_FOR_EACH(IMPL,2,B_SEQ)
+   BOOST_PP_SEQ_FOR_EACH(IMPL,5,C_SEQ)
+   BOOST_PP_SEQ_FOR_EACH(IMPL,9,D_SEQ)
+
+   void get_A_impl(const std::complex<double> [1], COLLIER_ARGS(0));
```

Comment:
During compilation I get the following warning at this point (and others):
~~~
warning: 'C1_impl' has C-linkage specified, but returns user-defined type 'std::complex<double>' which is incompatible with C [-Wreturn-type-c-linkage]
~~~
Since I'm not an expert on the Fortran/C/C++ interoperability, can you comment on the memory layout of the array of `std::complex<double>` and whether it is safe to map this to an array of complex numbers in Fortran?

I've only found some proposals, but I'm not aware of any guarantees.
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2004/n1589.html

Reply:
@Expander, yes, I was looking for the solution and it seems that this can also work: 
https://stackoverflow.com/questions/44180119/c-c-interop-between-complex-and-stdcomplex.

Reply:
Perfectly, thanks for the information!